Mathlib/Order/KonigLemma.lean

```diff
@@ -0,0 +1,160 @@
+/-
+Copyright (c) 2024 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson
+-/
+import Mathlib.Order.Cover
+import Mathlib.Order.WellFounded
+import Mathlib.Data.Set.Finite
+import Mathlib.Data.Finite.Set
+
+/-!
+# Kőnig's infinity lemma
+
+Kőnig's infinity lemma is most often stated as a graph theory result :
+every infinite, locally finite connected graph contains an infinite path.
+It has links to computability and proof theory, and
+is often applied when the graph is a tree whose vertices encode some other types of object.
+
+For applications of this form, it is convenient to state the lemma
+in terms of covers in a well-founded order rather than a graph;
+in this setting, the proof is almost trivial.
+In this file, we prove the lemma in this order-theoretic way,
+and also give a formulation that is convenient for applications in Ramsey theory.
+We defer the graph-theoretic version of the statement for future work.
+
+## Main Results
+
+* `exists_seq_covby_of_forall_covby_finite` : Kőnig's lemma for well-founded orders.
+
+* `exists_orderEmbedding_covby_of_forall_covby_finite` : Kőnig's lemma, where the sequence
+    is given as an `OrderEmbedding` instead of a function.
+
+* `exist_seq_forall_proj_of_forall_finite` : Kőnig's lemma applied to an order on a sigma-type
+  `(i : ℕ) × (α i)`. Useful for applications in Ramsey theory.
+
+## TODO
+
+Actually formulate the lemma as a statement about graphs.
+
+-/
+
+open Set
+
+variable {α : Type*} [PartialOrder α] [WellFoundedLT α] {a b : α}
+
+theorem exists_covby_of_lt (hab : a < b) : ∃ x, a ⋖ x ∧ x ≤ b := by
+  refine ⟨WellFounded.min wellFounded_lt (Ioc a b) ⟨b, hab,rfl.le⟩, ?_⟩
+  have hmem := (WellFounded.min_mem wellFounded_lt (Ioc a b) ⟨b, hab,rfl.le⟩)
+  exact ⟨⟨hmem.1,fun c hac hlt ↦ WellFounded.not_lt_min wellFounded_lt
+    (Ioc a b) ⟨b, hab,rfl.le⟩ ⟨hac, hlt.le.trans hmem.2⟩ hlt ⟩, hmem.2⟩
+
+theorem exists_covby_infinite_Ici_of_infinite_Ici (ha : (Ici a).Infinite)
+    (hfin : {x | a ⋖ x}.Finite) : ∃ b, a ⋖ b ∧ (Ici b).Infinite := by
+  by_contra! h
+  refine ((hfin.biUnion (t := Ici) (by simpa using h)).subset (fun b hb ↦ ?_)).not_infinite
+    (ha.diff (finite_singleton a))
+  obtain ⟨x, hax, hxb⟩ := exists_covby_of_lt (hb.1.lt_of_ne (Ne.symm hb.2))
+  exact mem_biUnion hax hxb
```

Comment:
I find it a bit weird to have these lemmas in a file whose name doesn't mention `Cover`. Maybe the whole file could be renamed to `Order.Cover.WellFounded` (and `Order.Cover` to `Order.Cover.Basic`)?

Reply:
I'm not sure about this - to me, covering is enough of a vanilla notion that it is fine for it to come up in files where it isn't the main focus, and conversely, the infinity lemma isn't really at heart about covering (its usual formulation makes no mention of orderings at all). Renaming the file would be misclassifying its contents.

Reply:
Sure, but `theorem exists_covby_of_lt (hab : a < b) : ∃ x, a ⋖ x ∧ x ≤ b` _really_ looks like an API lemma for `CovBy`

Reply:
See #14004